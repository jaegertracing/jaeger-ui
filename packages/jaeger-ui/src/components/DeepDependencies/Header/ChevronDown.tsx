// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { IoChevronDown } from 'react-icons/io5';

import './ChevronDown.css';

type TProps = {
  className?: string;
  style?: React.CSSProperties;
};

function ChevronDown(props: TProps) {
  const { className, style } = props;
  const cls = `Ddg--Header--ChevronDown ${className || ''}`;
  return <IoChevronDown className={cls} style={style} />;
}

export default React.memo(ChevronDown);
