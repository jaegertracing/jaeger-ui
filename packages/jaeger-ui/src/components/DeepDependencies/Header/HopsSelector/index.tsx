// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { memo } from 'react';

import { decode } from '../../../../model/ddg/visibility-codec';
import { ECheckedStatus, EDirection, TDdgDistanceToPathElems, THop } from '../../../../model/ddg/types';
import Selector from './Selector';

type TProps = {
  distanceToPathElems?: TDdgDistanceToPathElems;
  handleClick: (distance: number, direction: EDirection) => void;
  visEncoding?: string;
};

export default memo(function HopsSelector({ distanceToPathElems, handleClick, visEncoding }: TProps) {
  if (!distanceToPathElems) return <div />;

  const downstreamHops: THop[] = [];
  const upstreamHops: THop[] = [];
  let minFullDistance = 0;
  let minVisDistance = 0;
  let minVisDistanceFullness = ECheckedStatus.Empty;
  let maxFullDistance = 0;
  let maxVisDistance = 0;
  let maxVisDistanceFullness = ECheckedStatus.Empty;
  const visibleIndices = visEncoding && new Set(decode(visEncoding));

  distanceToPathElems.forEach((elems, distance) => {
    let fullness: ECheckedStatus;
    if (visibleIndices) {
      const visible = elems.filter(({ visibilityIdx }) => visibleIndices.has(visibilityIdx));
      if (visible.length === elems.length) {
        fullness = ECheckedStatus.Full;
      } else if (!visible.length) {
        fullness = ECheckedStatus.Empty;
      } else {
        fullness = ECheckedStatus.Partial;
      }
    } else {
      fullness = Math.abs(distance) <= 2 ? ECheckedStatus.Full : ECheckedStatus.Empty;
    }

    if (distance >= 0) downstreamHops.push({ distance, fullness });
    if (distance <= 0) upstreamHops.push({ distance, fullness });

    if (fullness !== ECheckedStatus.Empty) {
      if (distance >= maxVisDistance) {
        maxVisDistance = distance;
        maxVisDistanceFullness = fullness;
      }
      if (distance <= minVisDistance) {
        minVisDistance = distance;
        minVisDistanceFullness = fullness;
      }
    }

    if (fullness === ECheckedStatus.Full) {
      if (distance >= maxFullDistance) maxFullDistance = distance;
      if (distance <= minFullDistance) minFullDistance = distance;
    }
  });

  downstreamHops.sort(({ distance: a }, { distance: b }) => a - b);
  upstreamHops.sort(({ distance: a }, { distance: b }) => b - a);

  return (
    <div>
      <Selector
        direction={EDirection.Upstream}
        handleClick={handleClick}
        hops={upstreamHops}
        furthestDistance={minVisDistance}
        furthestFullDistance={minFullDistance}
        furthestFullness={minVisDistanceFullness}
      />
      <Selector
        direction={EDirection.Downstream}
        handleClick={handleClick}
        hops={downstreamHops}
        furthestDistance={maxVisDistance}
        furthestFullDistance={maxFullDistance}
        furthestFullness={maxVisDistanceFullness}
      />
    </div>
  );
});
