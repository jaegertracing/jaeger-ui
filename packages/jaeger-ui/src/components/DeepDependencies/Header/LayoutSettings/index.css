/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.Ddg--LayoutSettings {
  margin-right: 3em;
}
.Ddg--LayoutSettings--optionsTable > tbody > tr {
  vertical-align: top;
}

.Ddg--LayoutSettings--optionsTable > tbody > tr > td {
  padding: 0.75em;
}

.Ddg--LayoutSettings--optionsTable > tbody:not(:first-child) > tr:first-child > td {
  padding-top: 1.5em;
}

.Ddg--LayoutSettings--optionGroupTitle {
  color: #999;
  font-size: 0.95em;
}

.Ddg--LayoutSettings--option {
  display: flex;
  max-width: 500px;
}

.Ddg--LayoutSettings--optionNote {
  color: var(--tx-color-title);
}

.Ddg--LayoutSettings--optionDescription {
  margin-top: -0.2em;
  white-space: normal;
}
