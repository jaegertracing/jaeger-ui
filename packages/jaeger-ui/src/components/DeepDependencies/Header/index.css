/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DdgHeader {
  align-items: stretch;
  background: #fafafa;
  border-bottom: 1px solid #ddd;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  position: relative;
  z-index: 10;
}

.DdgHeader--controlHeader {
  align-items: center;
  background-color: #f6f6f6;
  display: flex;
  justify-content: space-between;
  padding-left: 1.25rem;
}

.DdgHeader--paramsHeader {
  align-items: center;
  background: #fafafa;
  border-bottom: 1px solid #eee;
  display: flex;
  flex: 1;
  flex-wrap: wrap;
  font-size: unset;
  margin: 0;
  padding: 0.75rem 1.25rem 0.75rem 1.25rem;
  position: relative;
}

.DdgHeader--findWrapper {
  display: flex;
  flex: 1 1 auto;
  justify-content: flex-end;
}

.DdgHeader--uiFind {
  align-items: center;
  background-color: #fff;
  border-left: 1px solid #eee;
  border-right: 1px solid #eee;
  color: #bbb;
  display: flex;
  flex-grow: 1;
  max-width: 400px;
  min-width: 200px;
  transition: max-width 0.5s;
}

.DdgHeader--uiFind:focus-within {
  max-width: 550px;
}

.DdgHeader--uiFindSearchIcon {
  margin: 0 0.25em;
  padding: 0.24em;
}

.DdgHeader--uiFindInput > input:focus,
.DdgHeader--uiFindInput > input {
  border: none;
  box-shadow: none;
  border-radius: 0px;
}

.DdgHeader--uiFindInfo {
  background: #ffffb8;
  border: 1px dashed #fadb14;
  border-radius: 4px;
  color: #ad8b00;
  margin-left: 0.25em;
  margin-right: 0.5em;
  padding: 0.27em 0.4em;
  stroke: #666;
  stroke-width: 1.2px;
  text-align: center;
  transition: width 0.5s;
  white-space: nowrap;
}

.DdgHeader--uiFindInfo--hidden {
  border-left: dashed var(--tx-color-muted) 1px;
  margin-left: 0.3em;
  padding-left: 0.3em;
}

.DdgHeader--uiFindInfo--icon {
  margin-bottom: 1.5px;
  margin-left: 0.2em;
}

.DdgHeader--uiFindInfo--tooltip {
  white-space: nowrap;
}
