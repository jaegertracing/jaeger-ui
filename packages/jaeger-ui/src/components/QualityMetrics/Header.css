/*
Copyright (c) 2020 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.QualityMetrics--Header {
  align-items: center;
  background: #fafafa;
  border-bottom: 1px solid #ddd;
  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: row;
  padding: 0.75rem 1.25rem 0.75rem 1.25rem;
  position: relative;
  z-index: 10;
}

.QualityMetrics--Header--LookbackLabel {
  color: var(--tx-color-muted);
  cursor: pointer;
  font-size: 1.5em;
  margin: 0 0.6em 0 0;
  outline: 4px solid transparent;
}

.QualityMetrics--Header--LookbackSuffix {
  color: var(--tx-color-muted);
  margin: 0 0 0 0.6em;
  outline: 4px solid transparent;
}
