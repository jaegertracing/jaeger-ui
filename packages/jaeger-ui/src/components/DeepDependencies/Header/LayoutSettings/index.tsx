// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Checkbox, Popover, Radio } from 'antd';
import { CheckboxChangeEvent } from 'antd/es/checkbox';
import { RadioChangeEvent } from 'antd/es/radio';

import settingsIcon from './settingsIcon';
import ChevronDown from '../ChevronDown';
import { EDdgDensity } from '../../../../model/ddg/types';
import { trackDensityChange, trackToggleShowOp } from '../../index.track';

import './index.css';

type TProps = {
  density: EDdgDensity;
  setDensity: (density: EDdgDensity) => void;
  showOperations: boolean;
  toggleShowOperations: (enable: boolean) => void;
};

const cssCls = (() => {
  const CLASSNAME_PREFIX = 'Ddg--LayoutSettings';
  return (namePart?: string) => (namePart ? `${CLASSNAME_PREFIX}--${namePart}` : CLASSNAME_PREFIX);
})();

// exported for tests
export const densityOptions = [
  {
    option: EDdgDensity.MostConcise,
    title: 'One node per resource',
    note: 'Most conscise',
    description:
      "This setting represents each resource one time in the graph, regardless of whether or not it's upstream or downstream of the focal node. This results in the most dense graph layout, possible.",
  },
  {
    option: EDdgDensity.UpstreamVsDownstream,
    title: 'Separate upstream from downstream nodes',
    description:
      'This setting differentiates upstream nodes from downstream nodes. And, within the each section, a given resource is represented by only one node.',
  },
  {
    option: EDdgDensity.OnePerLevel,
    title: 'One resource per level',
    description:
      'Represents each resource at most one time for any given level, or number of hops, from the focal node.',
  },
  {
    option: EDdgDensity.PreventPathEntanglement,
    title: 'Show paths to the focal node',
    description:
      'Each unique path to the focal node, both upstream and downstream, is represented. A resource (e.g. a service or service:operation) can be represented more than once if it is present in more than one path to the focal node.',
  },
  {
    option: EDdgDensity.ExternalVsInternal,
    title: 'Separate internal from external nodes',
    note: 'Most detailed',
    description:
      'This is similar to the setting, above, but this setting also differentiates internal from external nodes.',
  },
];

export default class LayoutSettings extends React.PureComponent<TProps> {
  private updateDensity = (event: RadioChangeEvent) => {
    const { density: prevDensity } = this.props;
    const { value: nextDensity } = event.target;
    if (prevDensity === nextDensity) return;
    trackDensityChange(prevDensity, nextDensity, densityOptions);
    this.props.setDensity(nextDensity);
  };

  private toggleShowOperations = (event: CheckboxChangeEvent) => {
    const { checked } = event.target;
    trackToggleShowOp(checked);
    this.props.toggleShowOperations(checked);
  };

  render() {
    const { density, showOperations } = this.props;
    const content = (
      <table className={cssCls('optionsTable')}>
        <tbody>
          <tr>
            <td className={cssCls('optionGroupTitle')}>Aggregations</td>
            <td>
              <Checkbox
                className={cssCls('option')}
                checked={showOperations}
                onChange={this.toggleShowOperations}
              >
                <div className={cssCls('optionDescription')}>
                  <h4>Show operations</h4>
                  <p>
                    Controls whether or not the operations are considered when creating nodes for the graph.
                  </p>
                </div>
              </Checkbox>
            </td>
          </tr>
        </tbody>
        <tbody>
          {densityOptions.map((config, i) => (
            <tr key={config.option}>
              {i === 0 && (
                <td className={cssCls('optionGroupTitle')} rowSpan={densityOptions.length}>
                  Graph density
                </td>
              )}
              <td>
                <Radio
                  className={cssCls('option')}
                  onChange={this.updateDensity}
                  value={config.option}
                  checked={density === config.option}
                >
                  <div className={cssCls('optionDescription')}>
                    <h4>{config.title}</h4>
                    <p>
                      {config.note && <em className={cssCls('optionNote')}>{config.note}</em>}
                      {config.note && ' â€“ '}
                      {config.description}
                    </p>
                  </div>
                </Radio>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    );
    return (
      <Popover arrowPointAtCenter content={content} placement="bottomLeft" title="Layout settings">
        <div className={cssCls()}>
          {settingsIcon}
          Layout
          <ChevronDown className="ub-ml2" />
        </div>
      </Popover>
    );
  }
}
