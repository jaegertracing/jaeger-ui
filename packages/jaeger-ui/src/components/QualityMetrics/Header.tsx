// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { InputNumber } from 'antd';
import _debounce from 'lodash/debounce';

import NameSelector from '../common/NameSelector';

import './Header.css';

type TProps = {
  lookback: number;
  service?: string;
  services?: string[] | null;
  setLookback: (lookback: number | string | undefined) => void;
  setService: (service: string) => void;
};

type TState = {
  ownInputValue: number | undefined;
};

export default class Header extends React.PureComponent<TProps, TState> {
  state: TState = {
    ownInputValue: undefined,
  };

  setLookback = _debounce((lookback: number | string | undefined) => {
    this.setState({ ownInputValue: undefined });
    this.props.setLookback(lookback);
  }, 350);

  handleInputChange = (value: string | number | undefined) => {
    if (typeof value === 'string') return;
    this.setState({ ownInputValue: value });
    this.setLookback(value);
  };

  render() {
    const { lookback, service, services, setService } = this.props;
    const { ownInputValue } = this.state;
    const lookbackValue = ownInputValue !== undefined ? ownInputValue : lookback;

    return (
      <header className="QualityMetrics--Header">
        <NameSelector
          label="Service"
          placeholder="Select a serviceâ€¦"
          value={service || null}
          setValue={setService}
          required
          options={services || []}
        />
        <label className="QualityMetrics--Header--LookbackLabel" htmlFor="inputNumber">
          Lookback:
        </label>
        <InputNumber id="inputNumber" onChange={this.handleInputChange} min={1} value={lookbackValue} />
        <span className="QualityMetrics--Header--LookbackSuffix">(in hours)</span>
      </header>
    );
  }
}
