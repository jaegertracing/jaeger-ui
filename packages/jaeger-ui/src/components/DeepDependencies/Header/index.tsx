// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Icon, Input, Tooltip } from 'antd';
import MdVisibility from 'react-icons/lib/md/visibility';
import MdVisibilityOff from 'react-icons/lib/md/visibility-off';

import HopsSelector from './HopsSelector';
import NameSelector from '../../common/NameSelector';
import LayoutSettings from './LayoutSettings';
import { trackFilter, trackHeaderSetOperation, trackShowMatches } from '../index.track';
import UiFindInput from '../../common/UiFindInput';
import { EDirection, TDdgDistanceToPathElems, EDdgDensity } from '../../../model/ddg/types';

import './index.css';

type TProps = {
  clearOperation: () => void;
  density: EDdgDensity;
  distanceToPathElems?: TDdgDistanceToPathElems;
  hiddenUiFindMatches?: Set<string>;
  operation?: string;
  operations: string[] | undefined;
  service?: string;
  services?: string[] | null;
  setDensity: (density: EDdgDensity) => void;
  setDistance: (distance: number, direction: EDirection) => void;
  setOperation: (operation: string) => void;
  setService: (service: string) => void;
  showOperations: boolean;
  showParameters?: boolean;
  showVertices: (vertexKeys: string[]) => void;
  toggleShowOperations: (enable: boolean) => void;
  uiFindCount: number | undefined;
  visEncoding?: string;
};
export default class Header extends React.PureComponent<TProps> {
  private _uiFindInput: React.RefObject<Input> = React.createRef();

  static defaultProps = {
    showParameters: true,
  };

  focusUiFindInput = () => {
    if (this._uiFindInput.current) {
      this._uiFindInput.current.focus();
    }
  };

  getUiFindInfo = () => {
    const { hiddenUiFindMatches, uiFindCount } = this.props;

    if (uiFindCount === undefined) return null;

    let hasHidden = false;
    let hiddenInfo: React.ReactNode = null;
    let tipText = uiFindCount ? 'All matches are visible' : 'No matches';
    if (hiddenUiFindMatches && hiddenUiFindMatches.size) {
      const { size } = hiddenUiFindMatches;
      hasHidden = true;
      tipText = `Click to view ${size} hidden match${size !== 1 ? 'es' : ''}`;
      hiddenInfo = (
        <span className="DdgHeader--uiFindInfo--hidden">
          {size}
          <MdVisibilityOff className="DdgHeader--uiFindInfo--icon" />
        </span>
      );
    }

    return (
      <Tooltip overlayClassName="DdgHeader--uiFindInfo--tooltip" placement="topRight" title={tipText}>
        {/* arbitrary span is necessary as Tooltip alters child's styling */}
        <span>
          <button
            className="DdgHeader--uiFindInfo"
            disabled={!hasHidden}
            onClick={this.handleInfoClick}
            type="button"
          >
            {uiFindCount}
            {(uiFindCount !== 0 || hasHidden) && <MdVisibility className="DdgHeader--uiFindInfo--icon" />}
            {hiddenInfo}
          </button>
        </span>
      </Tooltip>
    );
  };

  setOperation = (operation: string) => {
    trackHeaderSetOperation();
    this.props.setOperation(operation);
  };

  handleInfoClick = () => {
    trackShowMatches();
    const { hiddenUiFindMatches, showVertices } = this.props;
    if (hiddenUiFindMatches) showVertices(Array.from(hiddenUiFindMatches));
  };

  render() {
    const {
      clearOperation,
      density,
      distanceToPathElems,
      operation,
      operations,
      service,
      services,
      setDensity,
      setDistance,
      setService,
      showOperations,
      showParameters,
      toggleShowOperations,
      visEncoding,
    } = this.props;

    return (
      <header className="DdgHeader">
        {showParameters && (
          <div className="DdgHeader--paramsHeader">
            <NameSelector
              label="Service"
              placeholder="Select a service…"
              value={service || null}
              setValue={setService}
              required
              options={services || []}
            />
            {service && (
              <NameSelector
                clearValue={clearOperation}
                label="Operation"
                placeholder="Filter by operation…"
                value={operation || null}
                setValue={this.setOperation}
                options={operations || []}
              />
            )}
          </div>
        )}
        <div className="DdgHeader--controlHeader">
          <LayoutSettings
            density={density}
            setDensity={setDensity}
            showOperations={showOperations}
            toggleShowOperations={toggleShowOperations}
          />
          <HopsSelector
            distanceToPathElems={distanceToPathElems}
            handleClick={setDistance}
            visEncoding={visEncoding}
          />
          <div className="DdgHeader--findWrapper">
            <div className="DdgHeader--uiFind" role="button" onClick={this.focusUiFindInput}>
              <Icon className="DdgHeader--uiFindSearchIcon" type="search" />
              <UiFindInput
                allowClear
                forwardedRef={this._uiFindInput}
                inputProps={{ className: 'DdgHeader--uiFindInput' }}
                trackFindFunction={trackFilter}
              />
              {this.getUiFindInfo()}
            </div>
          </div>
        </div>
      </header>
    );
  }
}
