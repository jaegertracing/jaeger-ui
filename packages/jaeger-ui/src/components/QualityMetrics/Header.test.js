// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { InputNumber } from 'antd';
import debounceMock from 'lodash/debounce';

import Header from './Header';

jest.mock('lodash/debounce');

describe('Header', () => {
  const lookback = 4;
  const minProps = {
    lookback,
    setLookback: jest.fn(),
    setService: jest.fn(),
  };
  const service = 'test service';
  const props = {
    ...minProps,
    service,
    services: ['foo', 'bar', 'baz'],
  };
  let wrapper;
  let callDebouncedFn;
  let setLookbackSpy;

  beforeAll(() => {
    debounceMock.mockImplementation(fn => {
      setLookbackSpy = jest.fn((...args) => {
        callDebouncedFn = () => fn(...args);
      });
      return setLookbackSpy;
    });
  });

  beforeEach(() => {
    props.setLookback.mockReset();
    setLookbackSpy = undefined;
    wrapper = shallow(<Header {...props} />);
  });

  describe('rendering', () => {
    it('renders as expected with minimum props', () => {
      wrapper = shallow(<Header {...minProps} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders as expected with full props', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders props.lookback when state.ownInputValue is `undefined`', () => {
      expect(wrapper.find(InputNumber).prop('value')).toBe(lookback);
    });

    it('renders state.ownInputValue when it is not `undefined` regardless of props.lookback', () => {
      const ownInputValue = 27;
      wrapper.setState({ ownInputValue });
      expect(wrapper.find(InputNumber).prop('value')).toBe(ownInputValue);
    });
  });

  describe('setting lookback', () => {
    it('no-ops for string values', () => {
      wrapper.find(InputNumber).prop('onChange')('foo');
      expect(wrapper.state('ownInputValue')).toBe(undefined);
    });

    it('updates state with numeric value, then clears state and calls props.setLookback after debounce', () => {
      const value = 42;
      wrapper.find(InputNumber).prop('onChange')(value);

      expect(wrapper.state('ownInputValue')).toBe(value);
      expect(setLookbackSpy).toHaveBeenCalledWith(42);
      expect(props.setLookback).not.toHaveBeenCalled();

      callDebouncedFn();
      expect(wrapper.state('ownInputValue')).toBe(undefined);
      expect(props.setLookback).toHaveBeenCalledWith(42);
    });
  });
});
