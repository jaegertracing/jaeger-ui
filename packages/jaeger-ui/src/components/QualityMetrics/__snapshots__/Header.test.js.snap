// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header rendering renders as expected with full props 1`] = `
<header
  className="QualityMetrics--Header"
>
  <NameSelector
    label="Service"
    options={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    placeholder="Select a service…"
    required={true}
    setValue={[MockFunction]}
    value="test service"
  />
  <label
    className="QualityMetrics--Header--LookbackLabel"
    htmlFor="inputNumber"
  >
    Lookback:
  </label>
  <InputNumber
    id="inputNumber"
    min={1}
    onChange={[Function]}
    step={1}
    value={4}
  />
  <span
    className="QualityMetrics--Header--LookbackSuffix"
  >
    (in hours)
  </span>
</header>
`;

exports[`Header rendering renders as expected with minimum props 1`] = `
<header
  className="QualityMetrics--Header"
>
  <NameSelector
    label="Service"
    options={Array []}
    placeholder="Select a service…"
    required={true}
    setValue={[MockFunction]}
    value={null}
  />
  <label
    className="QualityMetrics--Header--LookbackLabel"
    htmlFor="inputNumber"
  >
    Lookback:
  </label>
  <InputNumber
    id="inputNumber"
    min={1}
    onChange={[Function]}
    step={1}
    value={4}
  />
  <span
    className="QualityMetrics--Header--LookbackSuffix"
  >
    (in hours)
  </span>
</header>
`;
